### PR TITLE
Merge strict and partial NBT ingredients

### LINKED ISSUE #328: partial nbt ingredient fails to match on numbers
**Minecraft Version:** 1.20.3

**NeoForge Version:** 20.3.3

**Steps to Reproduce:**
 1. Create a recipe using a partial nbt ingredient that checks if a repair able item has a repair cost of 1
 2. repair that item once
 3. try to use the item to craft the recipe you created
 4. notice that it does not work

**Description of issue:**
The partial nbt ingredient uses an `NbtPredicate` that uses `NbtUtils.compareNbt`, which requires both tags to be the same class this is an issue due to the tag for the stacks being created using `CompoundTag.CODEC` and how it works. As `CompoundTag.CODEC` uses `Dynamic.convert` and since its going from JSON to NBT it uses `JsonOps.convertTo`. `JsonOps` for numbers will convert to the type that is the smallest in size that can fit the number, (ie. 1 will always be converted to 1 byte). That means the item in-game will have NBT of `RepairCost` of 1 int while the `NbtPredicate` with have repair cost of 1 byte and they will never match.

**Example:**
Heres an example json recipe that will fail:
```json
{
  "type": "minecraft:crafting_shapeless",
  "ingredients": [
    {
      "type": "neoforge:partial_nbt",
      "item": "minecraft:diamond_axe",
      "tag": {
        "RepairCost": 1
      }
    }
  ],
  "result": {
    "item": "minecraft:bedrock"
  }
}
```

> Side note: But the `NbtPredicate` usage in `PartialNBTIngredient` is not really needed and can just be replaced with `NbtUtils.compareNbt` instead of creating a new instance of the predicate each time.

### PR DESCRIPTION
**Pull request by Matyrobbrt**
This PR merges the strict and partial ingredient into one `neoforge:nbt` one to avoid confusion and structure differences.
Fixes #328

### CODE REVIEW

**Review by Technici4n — COMMENTED** [review]
Would it be worth combining both strict and partial NBT ingredients under the same type, with an additional `"strict": true` for strict ingredients?

**Comment by ThatGravyBoat** [comment]
> Would it be worth combining both strict and partial NBT ingredients under the same type, with an additional `"strict": true` for strict ingredients?

I think combining them into one ingredient would be good as they are so similar, it would also be good as partial is not the most descriptive name unless you already knew about strict counterpart.

File: src/main/java/net/neoforged/neoforge/common/crafting/NBTIngredient.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    private static boolean compareStacksWithNBT(ItemStack left, ItemStack right) {
+        return left.getItem() == right.getItem() && NbtUtils.compareNbt(left.getTag(), right.getTag(), true);
+    }
+
+    /**
+     * Creates a new ingredient matching any item from the list, containing the given NBT
+     */
+    public static NBTIngredient of(boolean strict, CompoundTag nbt, ItemLike... items) {
+        return new NBTIngredient(Arrays.stream(items).map(ItemLike::asItem).collect(Collectors.toSet()), nbt, strict);
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/common/crafting/NBTIngredient.java:79]: Probably want to add `Preconditions.checkArgument(items.length > 0, "Must provide at least one item");`

Comment (Matyrobbrt) [reply src/main/java/net/neoforged/neoforge/common/crafting/NBTIngredient.java:79]: The constructor will already catch that.

**Review by Technici4n — APPROVED** [review]

File: src/main/java/net/neoforged/neoforge/common/crafting/NBTIngredient.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    /**
+     * Creates a new ingredient matching the given item, containing the given NBT
+     */
+    public static NBTIngredient of(boolean strict, ItemStack stack) {
```
Comment (KnightMiner) [inline src/main/java/net/neoforged/neoforge/common/crafting/NBTIngredient.java:85]: This really needs a javadoc explaining that non-strict matching may give unexpected results when using item stacks as they automatically set certain NBT keys (e.g. damagable items will be forced to match against a damage value of 0)

**Pull request merged**